### PR TITLE
[REVIEW] Adding null boolean handling for copy_if_else

### LINKED ISSUE #3856: [FEA] null boolean handling for copy_if_else
**Is your feature request related to a problem? Please describe.**
Currently `copy_if_else` throws a logic error if the boolean predicate vector contains any nulls.  When using `copy_if_else` to implement Spark SQL's `IF` or `CASE WHEN` statements, the boolean mask vector is typically the result of a comparison operation which can contain nulls.  It would be great if `copy_if_else` could handle the nulls.

**Describe the solution you'd like**
Spark SQL's `IF` and `CASE WHEN` are implemented like this pseudo-code:
```c++
  if (boolean_result == true) {
    return true_value;
  } else {
    return false_value;
  }
```
As a result, anything that is _not_ true is false, which means null entries are considered false.  Therefore I would either like `copy_if_else` to handle null boolean entries as if they were false or take an optional parameter to include that behavior.

**Describe alternatives you've considered**
This can be worked around by replacing the null entries in the boolean mask with `false` before calling `copy_if_else`, but it requires manifesting an intermediate boolean vector.  It would be more efficient if `copy_if_else` handled null booleans directly.

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
Adding null boolean handling for copy_if_else.
close #3856

### CODE REVIEW

File: cpp/src/copying/copy.cu [hunk 1]
```diff
@@ -132,14 +131,26 @@ std::unique_ptr<column> copy_if_else( Left const& lhs,
...
    column_device_view bool_mask_device = *bool_mask_device_p;                                    
-   
-   auto filter = [bool_mask_device] __device__ (cudf::size_type i) { return bool_mask_device.element<cudf::experimental::bool8>(i); };
-   return cudf::experimental::type_dispatcher(lhs.type(),                                             
-                                             copy_if_else_functor{},
-                                             lhs, rhs, boolean_mask.size(),
-                                             left_nullable, right_nullable,
-                                             filter,
-                                             mr, stream);
+
```
Comment (nvdbaranec) [inline cpp/src/copying/copy.cu]: This works but see Jake's comment here.  https://github.com/rapidsai/cudf/issues/3856#issuecomment-576949361

File: cpp/src/copying/copy.cu [hunk 2]
```diff
@@ -132,14 +131,26 @@ std::unique_ptr<column> copy_if_else( Left const& lhs,
...
-   
-   auto filter = [bool_mask_device] __device__ (cudf::size_type i) { return bool_mask_device.element<cudf::experimental::bool8>(i); };
-   return cudf::experimental::type_dispatcher(lhs.type(),                                             
-                                             copy_if_else_functor{},
-                                             lhs, rhs, boolean_mask.size(),
-                                             left_nullable, right_nullable,
-                                             filter,
-                                             mr, stream);
+  
+   if (boolean_mask.nullable() and boolean_mask.has_nulls()) {
```
Comment (jrhemstad) [inline cpp/src/copying/copy.cu]:
```suggestion
   if (boolean_mask.has_nulls()) { 
```

File: cpp/src/copying/copy.cu (hunk 1 above)
Comment (rgsl888prabhu) [reply cpp/src/copying/copy.cu]: I tried it, but there were limitations due to the iterator form, so I had to switch back to this.

**Comment by nvdbaranec** [comment]
:+1:

File: cpp/src/copying/copy.cu (hunk 1 above)
Comment (nvdbaranec) [reply cpp/src/copying/copy.cu]: :+1:

**Comment by rgsl888prabhu** [comment]
@davidwendt @ChuckHastings

File: cpp/src/copying/copy.cu (hunk 2 above)
Comment (davidwendt) [inline cpp/src/copying/copy.cu]: You should only need to check for `has_nulls()` here. If there are nulls that means it also is nullable.

Comment (rgsl888prabhu) [reply cpp/src/copying/copy.cu]: has_nulls() directly starts null_count rather than looking if there is a null mask or not.

Comment (davidwendt) [reply cpp/src/copying/copy.cu]: Well `null_count` cannot be `>0` without a `null_mask`.

Comment (rgsl888prabhu) [reply cpp/src/copying/copy.cu]: I was just concerned about multiple calls, but it shouldn't be much of a burden.

**Comment by rgsl888prabhu** [comment]
rerun tests

File: cpp/include/cudf/copying.hpp [hunk 3]
```diff
@@ -486,12 +486,11 @@ std::vector<contiguous_split_result> contiguous_split(cudf::table_view const& in
  *          
  * @throws cudf::logic_error if lhs and rhs are not of the same type
  * @throws cudf::logic_error if lhs and rhs are not of the same length 
- * @throws cudf::logic_error if boolean_mask contains nulls
  * @throws cudf::logic_error if boolean mask is not of type bool8
  * @throws cudf::logic_error if boolean mask is not of the same length as lhs and rhs  
  * @param[in] left-hand column_view
  * @param[in] right-hand column_view
- * @param[in] Non-nullable column of `BOOL8` elements that control selection from `lhs` or `rhs`
+ * @param[in] column of `BOOL8` elements that control selection from `lhs` or `rhs`
```
Comment (jrhemstad) [inline cpp/include/cudf/copying.hpp]: Should probably add documentation that describes how null values are treated as false.

**Comment by rgsl888prabhu** [comment]
rerun tests

**Pull request merged**